### PR TITLE
Refactor for...in to for...of

### PR DESCRIPTION
**Pull request by otaku**
## Original Problem
for...in will iterate inherited enumerable properties.  Other nodes added properties to Array and were being iterated as well causing strange errors.

## Solution
The following refactors were applied:

- for...in was changed to for...of.  In places where Typescript complains, a guard was added.
- added a radix of 10 to all parseInt calls
- removed unnecessary escape characters in regular expression strings
- removed unnecessary semicolons in class function declaration
- changed Typescript namespace to ES2015 class

### CODE REVIEW

File: src/lib/controller/RemoteController.ts [hunk 1]
```diff
@@ -1066,11 +1064,13 @@ export class RemoteController extends EventEmitter implements SerializableContro
 
     private handleDataStreamConnectionClosed(connection: DataStreamConnection): void {
         for (const targetIdentifier in this.dataStreamConnections) {
```
Comment (Supereg) [inline src/lib/controller/RemoteController.ts:1066]: Can this also be transformed to a `for ... of`? 
Using `for (const [targetIdentifier, connection0] of Object.entries(this.dataStreamConnections))` so you are able to do the `delete`?

**Review by Supereg — COMMENTED** [review]
Appart from the one comment PR looks good. Thanks for the contribution 👍 🎉 
Maybe you can quickly elaborate on that specific decision, otherwise happy to merge.

**Comment by Supereg** [comment]
Did not have a chance yet to read through #872. Can this issu? be closed once this PR is merged?

File: src/lib/Accessory.ts [hunk 2]
```diff
@@ -437,8 +437,7 @@ export class Accessory extends EventEmitter {
         : serviceParam;
 
     // check for UUID+subtype conflict
-    for (let index in this.services) {
-      const existing = this.services[index];
+    for (let existing of this.services) {
```
Comment (Shaquu) [inline src/lib/Accessory.ts:440]: Can it be const?

**Review by Shaquu — COMMENTED** [review]
One small comment from my side also.
I noticed one TODO, is it possible to resolve in this PR?

**Comment by Shaquu** [comment]
@Supereg It should resolve issue from what first reporter said.

NRCHKB/node-red-contrib-homekit-bridged#359

File: src/lib/controller/RemoteController.ts (hunk 1 above)
Comment (otaku) [reply src/lib/controller/RemoteController.ts]: The issue is with Typescript actually.  Object.keys and Object.entries both return key as string.

As an alternative, I converted Record into a Map.  I was considering applying the same change to other Record variables as well if this is acceptable.

File: src/lib/controller/RemoteController.ts [hunk 3]
```diff
@@ -58,7 +58,7 @@ const enum SupportedButtonConfigurationTypes {
     BUTTON_TYPE = 0x02
 }
 
-export const enum ButtonType {
```
Comment (otaku) [inline src/lib/controller/RemoteController.ts:61]: I removed the `const` here to allow the removal of a `@ts-ignore`

File: src/lib/controller/RemoteController.ts [hunk 4]
```diff
@@ -58,7 +58,7 @@ const enum SupportedButtonConfigurationTypes {
...
+export enum ButtonType {
```
Comment (Supereg) [inline src/lib/controller/RemoteController.ts:61]: I would like to keep this a `const enum` If possible. This has some implications on homebridge plugins, as using const enums, they can easily just use type imports. For JavaScript compatibility we also have turned on the „preserve const enums“ transpiler option.
I know this makes usage of this enum a bit painful and requires some `@ts-ignore` or `@ts-expect-error` in some places.

File: src/lib/controller/RemoteController.ts [hunk 5]
```diff
@@ -1063,14 +1071,11 @@ export class RemoteController extends EventEmitter implements SerializableContro
...
     private handleDataStreamConnectionClosed(connection: DataStreamConnection): void {
-        for (const targetIdentifier in this.dataStreamConnections) {
-            if (Object.prototype.hasOwnProperty.call(this.dataStreamConnections, targetIdentifier)) {
-                const connection0 = this.dataStreamConnections[targetIdentifier];
-                if (connection === connection0) {
-                    debug("HDS connection disconnected for targetIdentifier %s", targetIdentifier);
-                    delete this.dataStreamConnections[targetIdentifier];
-                    break;
-                }
+        for (const [ targetIdentifier, connection0 ] of this.dataStreamConnections) {
```
Comment (Supereg) [inline src/lib/controller/RemoteController.ts:1074]: Like the migration to `Map`, makes this a lot simpler 👍

File: src/lib/controller/RemoteController.ts (hunk 3 above)
Comment (Supereg) [reply src/lib/controller/RemoteController.ts:61]: See https://github.com/homebridge/HAP-NodeJS/pull/879#discussion_r575848556

**Comment by otaku** [comment]
Would you like more refactoring applied to this PR or opened in a new PR?  I'd be targeting Record to Map specifically.

File: src/lib/controller/RemoteController.ts (hunk 1 above)
Comment (Supereg) [reply src/lib/controller/RemoteController.ts]: Ah yeah, makes sense.

If you like, you can migrate that in other places as well, though maybe doing this in a separate PR? So we can merge this, as it seemingly solves some important issues?

Is there any disadvantage to regular objects/Records? Like performance wise? I‘m not really educated on Map vs Record.

**Review by Supereg — DISMISSED** [review]

**Comment by otaku** [comment]
I noticed I serialized a Map, which JSON.stringify will fail to do.

File: src/lib/controller/RemoteController.ts [hunk 6]
```diff
@@ -346,7 +346,7 @@ interface RemoteControllerServiceMap extends ControllerServiceMap {
 
 interface SerializedControllerState {
     activeIdentifier: number,
-    targetConfigurations: Map<number, TargetConfiguration>;
+    targetConfigurations: [number, TargetConfiguration][];
```
Comment (Supereg) [inline src/lib/controller/RemoteController.ts:349]: `targetConfigurations` was a Record previously right? Is this still compatible with previous created storage entries?

**Comment by otaku** [comment]
Thank you, I restored the type and serialize / deserialize from Record to Map instead.  Let me know if I should just rip out Map instead? Though it may still be better as it's less prone to enumerating inherited properties.

**Review by Supereg — APPROVED** [review]
Even if serialization and deserialization isn't as clean, I think having Map is nonetheless an improvement 👍 
Thanks for the contribution.

**Comment by Supereg** [comment]
can we merge the PR or is something still missing?

**Comment by otaku** [comment]
Nothing else is missing. This PR is complete. Thank you!

**Pull request merged**